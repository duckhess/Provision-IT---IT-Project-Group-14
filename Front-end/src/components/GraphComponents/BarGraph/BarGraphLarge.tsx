import { ResponsiveContainer, BarChart, CartesianGrid, Bar, XAxis, YAxis, Tooltip, Legend } from 'recharts';
import type { Dataset } from '../../Types/Types';

interface GraphProps {
  datasets: Dataset[]; 
  mergedSets: Dataset[];
  yLabel: String
  title: String
}

const colors = ["#8884d8", "#82ca9d", "#ffc658", "#ff7300"];

const BarGraphLarge = ({ datasets, mergedSets, yLabel, title }: GraphProps) => {
  return (

    
<div className="flex flex-col items-start w-[100%] h-[800px] bg-gray-100 rounded-lg shadow p-4">
  <div className ="px-4 w-full">
    <h2 className='text-black text-xl font-bold border-b mb-4 inline-block'>
      {title}
    </h2>
  </div>
  <div className="w-full h-full">
    <ResponsiveContainer width="100%" height="100%">
      <BarChart data={mergedSets}>
        <CartesianGrid stroke="#ccc" />
        <XAxis dataKey="x" />
        <YAxis tickFormatter={(v) => `${v}${yLabel}`} />
        <Tooltip />
        <Legend />
        {datasets.map((ds, i) => (
          <Bar
            key={i}
            type="monotone"
            dataKey={ds.name}
            fill={colors[i % colors.length]}
          />
        ))}
      </BarChart>
    </ResponsiveContainer>
  </div>
</div>
  )
}

export default BarGraphLarge