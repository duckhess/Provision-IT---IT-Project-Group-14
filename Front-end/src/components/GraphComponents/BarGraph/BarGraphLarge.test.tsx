import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import BarGraphLarge from "./BarGraphLarge";
import type { Dataset } from "../../Types/Types";

// Mock Recharts components to render HTML instead of SVG
vi.mock("recharts", async (importOriginal) => {
  const original: any = await importOriginal();
  return {
    ...original,
    ResponsiveContainer: ({ children }: any) => <div>{children}</div>,
    BarChart: ({ children }: any) => <div>{children}</div>,
    CartesianGrid: () => <div />,
    XAxis: () => <div />,
    YAxis: () => <div />,
    Tooltip: () => <div />,
    Legend: ({ payload }: any) => (
      <div data-testid="mock-legend">
        {/* Mock the legend items for testing */}
        <span>Revenue</span>
        <span>Expenses</span>
      </div>
    ),
    Bar: ({ dataKey }: any) => <div data-testid={`bar-${dataKey}`} />,
  };
});

describe("BarGraphLarge", () => {
  const mockDatasets: Dataset[] = [
    {
      name: "Revenue",
      data: [
        { x: "Jan", y: 1000 },
        { x: "Feb", y: 1200 },
      ],
      metric: "Income Statements",
      unit: "Ratio",
    },
    {
      name: "Expenses",
      data: [
        { x: "Jan", y: 500 },
        { x: "Feb", y: 600 },
      ],
      metric: "Income Statements",
      unit: "Ratio",
    },
  ];

  const mockMergedSets: Array<{ [key: string]: number | string }> = [
    { x: "Jan", Revenue: 1000, Expenses: 500 },
    { x: "Feb", Revenue: 1200, Expenses: 600 },
  ];

  it("renders the chart title", () => {
    render(
      <BarGraphLarge
        datasets={mockDatasets}
        mergedSets={mockMergedSets}
        yLabel="Ratio"
        title="Income Statements"
      />
    );

    expect(screen.getByText("Income Statements")).toBeInTheDocument();
  });

  it("renders the bars with correct dataKeys", () => {
    render(
      <BarGraphLarge
        datasets={mockDatasets}
        mergedSets={mockMergedSets}
        yLabel="Ratio"
        title="Income Statements"
      />
    );

    // Check that each Bar component is rendered with the correct dataKey
    mockDatasets.forEach(ds => {
      expect(screen.getByTestId(`bar-${ds.name}`)).toBeInTheDocument();
    });
  });

  it("renders legend items", () => {
    render(
      <BarGraphLarge
        datasets={mockDatasets}
        mergedSets={mockMergedSets}
        yLabel="Ratio"
        title="Income Statements"
      />
    );

    expect(screen.getByText("Revenue")).toBeInTheDocument();
    expect(screen.getByText("Expenses")).toBeInTheDocument();
  });

  it("renders without crashing and matches snapshot", () => {
    const { container } = render(
      <BarGraphLarge
        datasets={mockDatasets}
        mergedSets={mockMergedSets}
        yLabel="Ratio"
        title="Income Statements"
      />
    );

    expect(container).toMatchSnapshot();
  });
});
