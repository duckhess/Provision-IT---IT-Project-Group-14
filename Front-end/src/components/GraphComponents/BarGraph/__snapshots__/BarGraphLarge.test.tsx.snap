// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BarGraphLarge > renders without crashing and matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col items-start w-[100%] h-[800px] bg-gray-100 rounded-lg shadow p-4"
  >
    <div
      class="px-4 w-full"
    >
      <h2
        class="text-black text-xl font-bold border-b mb-4 inline-block"
      >
        Income Statements
      </h2>
    </div>
    <div
      class="w-full h-full"
    >
      <div>
        <div>
          <div />
          <div />
          <div />
          <div />
          <div
            data-testid="mock-legend"
          >
            <span>
              Revenue
            </span>
            <span>
              Expenses
            </span>
          </div>
          <div
            data-testid="bar-Revenue"
          />
          <div
            data-testid="bar-Expenses"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
